### ISSUE TITLE
OutOfMemoryError PermGen space and hanging builds

### ISSUE THREAD

**Issue by cameocoder** [robolectric/robolectric#1700]
I have a suite consisting of almost 2000 tests that work with Robolectric 2.4.  I have started updating the suite to work with 3.0-rc2 but I can only run about ~700 tests before they fail with an OutOfMemoryError or hang.  It seems that once I reach a certain size, the tests will fail.  For example I will run X tests successfully, then I will add a few more test classes and the suite will fail.  If I leave the newly added classes and remove some of the classes known to be successful, I can get a successful build again.

The conversion for the majority of tests was just a matter of using the RobolectricGradleTestRunner,  adding @Config line and fixing references to Robolectric classes that were moved.  Continuing to use RobolectricTestRunner instead of RobolectricGradleTestRunner fails in similar ways.

The following jvmargs are set in the gradle.properties file.  Increasing the memory options does not help:
org.gradle.jvmargs=-Xmx1024m -XX:MaxPermSize=512m -XX:+HeapDumpOnOutOfMemoryError -Dfile.encoding=UTF-8

There are a few failure scenarios I have seen:
- Test execution hang with no output
- "OutOfMemoryError: PermGen space" error followed by gradle stacktrace
- One of the above or just test failure followed by a different class loading error each time:
    [ERROR] couldn't load roboguice.activity.RoboActivity in org.robolectric.internal.bytecode.InstrumentingClassLoader@4d0ed336

    [ERROR] couldn't load android.content.Context in org.robolectric.internal.bytecode.InstrumentingClassLoader@3e1cbf2e

    [ERROR] couldn't load org.bouncycastle.jce.provider.BouncyCastleProvider in
org.robolectric.internal.bytecode.InstrumentingClassLoader@6243e098

In all cases, just removing a few test classes will allow the tests to succeed.

Given I have a suite of testcases that work with 2.4 and work with 3.0-rc2 in batches of ~700 tests, It seems to point to something in the 3.0 changes that are causing these OOM errors.

**Comment by cameocoder** [comment]
I notices something interesting when running my Robolectric upgrade branch on another machine.  The number of tests that will execute before the OOM error depends on the number of parallel process used to run the tests.

For example, running the tests serially will run about ~400 tests.  But we typically run tests in parallel with the following options:
subprojects {
    tasks.withType(Test) {
        maxParallelForks = Runtime.runtime.availableProcessors()
    }
}

With maxParallelForks set to 2, ~700 tests will run.
With maxParallelForks set to 4 ~1200 tests will run.

**Comment by jongerrish** [comment]
I've been seeing similar problems sporadically.

What are your jvm permgen settings?
Did you specify a particular version of Android SDK to use?
What test runner(s) are you using?

In Robolectric tests, especially where liberal use of Mockito is involved requires a lot of permgen space due to all the generated classes. Probably this is exasperated by newer and bigger android sdks. If you just moved from 2.4 to 3.0 without any config changes then I think you've also moved from emulateSdk = 18 to emulateSdk = 21 implicitly.

Quick fixes would be to:-
1) Bump your max permgen setting
2) Set Robolectric config to use emulateSdk = 18 for all your tests (same as the default in 2.4)

The more involved was would be to check where this permgen space is going, its possible there is a leak... I followed these instructions to get a better understanding of possible issues and how to diagnose.

https://cdivilly.wordpress.com/2012/04/23/permgen-memory-leak/ 

When I checked yesterday, I found that one problem is when you are using multiple test runner instances. Each one of them holds its own reference to a current and previous (cached) SdkEnvironment which in turn holds a reference to the instrumenting classloader which contains class definitions for everything it has loaded. Too many of these and it soon maxes out perm gen. You can test this by creating a custom test runner for each of your tests.

@erd this is one of the things that I was chatting to you about yesterday

The solution I think is to have a static (VM + test runner wide cache of all SdkEnvironments) which means RobolectricTestRunner classes would share this cache. It is probably better anyway these days as there are more android versions to test against, more and more tests will have different tests for different api levels.

I'll try and submit a PR for that today but if I were you I would profile your tests anyway, start by querying for 

select x from org.robolectric.internal.SdkEnvironment x

How many are there? What objects are holding a reference to it?

**Comment by mag** [comment]
I have been working on a large app that was having these problems as well.

We recently switched over to Java 8, which eliminates perm gen altogether,
and these problems went away. We are on Robolectric 2.4.

Mike

**Comment by cameocoder** [comment]
The only place we specify permgen setting is in our gradle.properties file:
org.gradle.jvmargs=-Xmx1024m -XX:MaxPermSize=512m -XX:+HeapDumpOnOutOfMemoryError -Dfile.encoding=UTF-8

Updating those settings did not seem to impact the OOM errors.

I currently use "emulateSdk = 19" for all tests.  We currently have the following config:
        minSdkVersion 14
        targetSdkVersion 19

I actually managed to get past the OOM errors by adding the following to our top level build.gradle file:
subprojects {
    tasks.withType(Test) {
        maxParallelForks = Runtime.runtime.availableProcessors()
        maxHeapSize = "512m"
    }
}

I can replace maxHeapSize with jvmArgs '-XX:MaxPermSize=256m' and that also allows the tests to pass.

**Comment by cameocoder** [comment]
I updated to JDK8 on one machine and it also seems to resolve the OOM errors as @mag pointed out.

I ran a heap dump and there were no instances of org.robolectric.internal.SdkEnvironment.

The third top reference was for org.robolectric.res.TypedResource and a lot of the top instances in general were in the org.robolectric.res package.

I figured out a third solution to the OOM errors which may also point to the cause.  When I started converting our code from 2.4 to 3.0 I added a Config line after every RunWith(RobolectricTestRunner) annotation.  It turns out we have some test classes that have nested classes which looked like this:
```java
@RunWith(Enclosed.class)
public class FooTest {
    @RunWith(RobolectricTestRunner.class)
    @Config(constants = BuildConfig.class, emulateSdk = 19)
    public static class Foo1Test extends FooTest {
...
    }

    @RunWith(RobolectricTestRunner.class)
    @Config(constants = BuildConfig.class, emulateSdk = 19)
    public static class Foo2Test extends FooTest {
...
    }
}
```
As part of my cleanup tasks I removed the multiple instances of the Config annotation and moved it to the parent class.  I didn't even have to fix them all before the tests started running without OOM errors.

**Comment by kageiit** [comment]
I can confirm updating to JDK 8 also resolves the OOM issues when running a large suite of tests. Is there any way to fix this when running on JDK 7?

**Comment by jongerrish** [comment]
I'll profile this tonight if I can, feels like there is a leak somewhere.

**Comment by jongerrish** [comment]
I _think_ the cause of this is due to BackgroundThread leaking the InstrumentingClassloader through its contextClassloader field, it has a static singleton that probably needs resetting between test runs and the looper should probably be shut down also.

**Comment by jongerrish** [comment]
There seem to be a couple of other spots:-

1) ViewRootImpl.runQueues 
2) A thread in the sql connection code
3) A case of AppContext

I'll take a look at these next.

I think the BackgroundThread leak might have been a big one, since we started to notice permgen issues on api 19+ ....also, it was the cause of two leaks for the Robolectric test suite, we should fix the other leaks too.

The original fix I submitted (https://github.com/robolectric/robolectric/pull/1724) which intended to cache the InstrumentingClassloader doesn't work because Robolectric classloaders are unique across SdkConfig + InstrumentedClassloaderConfig. Unfortunately Robolectrics test suites have multiple test runners all with their own InstrumentedClassloaderConfig which makes things more complicated.... originally I was supposing that the leaks wouldn't be so problematic if we were just reusing a unique and cached set of InstrumentingClassloaders one per SdkConfig (16, 17, 18, 19, 21).... as it is this is really just an issue where you are creating many classloaders, as is the case in the Robolectric test suite where there are many different test runners with their own class loader configuration.

**Comment by lexs** [comment]
While it's true that runners can override the `InstrumentedClassloaderConfig` used this doesn't seem to be done outside of internal Robolectric tests? So I still think it's valid to have a global cache for these as long as `config.getClass() == InstrumentedClassloaderConfig.class`.

**Comment by jongerrish** [comment]
Yes that's right :-)

**Comment by jongerrish** [comment]
Hey @lexs + @erd would you mind taking a look at https://github.com/robolectric/robolectric/pull/1746 - this is the caching of SdkEnvironments we have been talking about.

**Comment by jaredsburrows** [comment]
@cameocoder Have you tried upgrading to Java 8? Can this be closed?

**Comment by erd** [comment]
We did some work with 3.0 to make this better, so I'm going to close this for now. We can revisit in the future if it continues to crop up.

**Comment by andrewlord1990** [comment]
I have just upgraded our project to Robolectric 3.0. However, we still get these OOM errors PermGen Space after about 450 of our 1000 tests have ran. I have upped the JVM args as mentioned above and laptop has 16GB RAM, still can't run them all.
How come this issue was closed?

**Comment by mttmllns** [comment]
Just upgraded to 3.0.  Seeing OOM PermGen errors on command line.  Running the tests from Android Studio works fine after editing the run configuration to use -XX:MaxPermSize=512M.  But doing the same for gradle.properties doesn't work for command line.  Even tried something obscene:
```
org.gradle.jvmargs=-Xms3g -Xmx3g -XX:MaxPermSize=3g -XX:+HeapDumpOnOutOfMemoryError
```
Any suggestions for troubleshooting?

**Comment by kageiit** [comment]
In your `build.gradle`: 
```groovy
tasks.withType(Test) {
  jvmArgs '-Xms3g -Xmx3g -XX:MaxPermSize=3g -XX:+HeapDumpOnOutOfMemoryError'
}
```

Since gradle runs tests in a separate vm.

**Comment by mag** [comment]
This was mentioned earlier in the thread but I'll mention it again: try
using Java 8. My project was having lots of hanging builds and permgen
errors, but Java 8 made them all go away.

Mike

**Comment by jaredsburrows** [comment]
@mag  +1

**Comment by mttmllns** [comment]
Thanks @kageiit, that works perfectly.  I didn't want to force all our developers to upgrade to Java 8 just yet.

```groovy
tasks.withType(Test) {
	jvmArgs '-XX:MaxPermSize=512m'
}
```

**Comment by JavierSP1209** [comment]
@erd Hi! I'm having also memory issues after upgrading to Robolectric 3.0, I have a suite consisting of almost 1500 tests that worked with Robolectric 2.4, but after upgrading to 3.0 some tests just hangs or throws OutOfMemoryExceptions. I'm using Java 8 and non of the jvmArgs seems to help, I made some Memory Dumps and found out that the InstrumentingClassLoader instances are not been garbage collected after the test its done, for example this is the output of the memory leak analysis:

850 instances of "org.robolectric.shadows.ShadowApplication$1", loaded by "org.robolectric.internal.bytecode.InstrumentingClassLoader @ 0x6c08572a0" occupy 1,036,212,944 (35.43%) 

528,488 instances of "android.app.LoadedApk$ReceiverDispatcher", loaded by "org.robolectric.internal.bytecode.InstrumentingClassLoader @ 0x6c08572a0" occupy 1,032,349,744 (35.30%) bytes. 

These memory dump was created after running the first 900 tests, so as you can see we almost have an instance of ShadowApplication for each test thats its not been GC.

Javier.

**Comment by hidroh** [comment]
@JavierSP1209 so how do you workaround this?

**Comment by JavierSP1209** [comment]
@hidroh The work around was to create multiple forks so the tests could run in different JVMs in this way I don't reach the heap size limit, I'm still eaten up all my RAM memory when I run the tests but at least they finish. When running form AS, I add the -Xms2G -Xmx6G jvm args on the running configs for JUnit

**Comment by danlegion** [comment]
@kageiit +1

**Comment by andrewlord1990** [comment]
Did upgrading to JDK8 not fix the issue for you (as mentioned by earlier comments)? As soon as we upgraded to JDK8 all of the tests started running successfully.

**Comment by intirix** [comment]
Upgrading to JDK8 did not fix the issue for me.  I'm migrating from 2.3 to 3.0 and API 18 to 21.  JProfiler is showing a lot of Strings in memory that get referenced by ResBundles.  Most of the ResBundles are referenced from PackageResourceLoader.  I am not using a custom test runner and all Config annotations are the same.

**Comment by mag** [comment]
#2068 may be your actual problem. There are some workarounds in there that helped tamp-down the memory leak in my large-ish suite, but there is more work to be done.

**Comment by jaredsburrows** [comment]
@andrewlord1990 Java 8 did for me

@intirix On JDK7 I have to specify `jvmargs`. What do you mean `Config` annotations are the same? From 2.3 to 3.0, there are major changes. Also, in your `tearDown` methods, are you destroying your fragments and activities?

**Comment by intirix** [comment]
@mag I put that in but it didn't seem to help things.  I will continue to play around with my teardown to clean things up

@jaredsburrows Every test class has the same exact @Config(constants=BuildConfig.class, sdk=21, application=MyCustomApp.class) line.  Some tests do destroy the Activities they create but some do not.  I will try to go through the unit tests that don't destroy their Activities and destroy them.

**Comment by jaredsburrows** [comment]
@intirix Thanks for the quick reply. One thing I started doing is for all fragments and activities, I setup an `activitycontroller` and `destroy` each fragment and activity in the `teardown`.

**Comment by intirix** [comment]
@jaredsburrows thanks for the suggestion.  2/3 of my tests already destroyed the activity.  I updated the last 1/3 to destroy the activity.  That helped out a lot, but I still ran out of memory.  I don't explicitly create nor destroy fragments.  I was letting the activity lifecycle do that.  I'm still playing around with it, but I don't want to take over this issue's discussion thread.

**Comment by jaredsburrows** [comment]
@intirix No you are being helpful be giving feedback :) :+1: 

- When you say `2/3 tests`, do you mean `(2/3) * 1000`? How many tests or after how many tests do you see this error?
- Can you show examples of the tests that ran that causes this error?

One thing that has helped me is to abstract POJOs from Robolectric and unit test without RobolectricTestRunner and use Junit test runner.

**Comment by intirix** [comment]
I have 186 tests.  About 50 of them didn't destroy the activity.  Before I made the change, only about 10 test cases ran before I ran out of memory.  Now, I get to 150 before running out of memory.  Each individual test case runs just fine.  I will try to get approval to share test cases.

I monitored the JVM using JConsole and noticed that the activities that have the largest number of views tend to be the ones that create the most memory.  I have two pages that have a ginormous number of views and inflations.  I'm going to spend some time optimizing those tests.

**Comment by jaredsburrows** [comment]
@intirix Thanks for the feedback again. It sounds like it helps. Views definitely push up memory. Have you tried allocating more memory?

**Comment by kageiit** [comment]
I have encountered this issue before. Using jvisualvm, I obtained heap dumps and profiled them using Eclipse heap analyzer. The `android.view.WindowManagerGlobal` class (which is hidden) was keeping references to all the inflated views and that caused OOMs eventually.

It was solved when I invoked the `closeAll` static method on that class via reflection in my base class after each test in in @After block. This might be a good optimization to fold into robolectric's base test runner itself

**Comment by jaredsburrows** [comment]
@kageiit Can you share your code in your `tearDown`?

**Comment by intirix** [comment]
@jaredsburrows I have increased my heap to 2G which is definitely helping.  Still trying to narrow down the issue.  The one set of tests that seem to cause heap usage to spike are calling this block of code:

		ShadowListView slv = Shadows.shadowOf(listView);
		slv.populateItems();

@kageiit I did something similar, but #2068 had alternative code due to a bug in closeAll.  Here is what is in my teardown:

		// https://github.com/robolectric/robolectric/pull/1741
		final Class<?> btclass = Class.forName("com.android.internal.os.BackgroundThread");
		Object backgroundThreadSingleton = ReflectionHelpers.getStaticField(btclass,"sInstance");
		if (backgroundThreadSingleton!=null) {
			btclass.getMethod("quit").invoke(backgroundThreadSingleton);
			ReflectionHelpers.setStaticField(btclass, "sInstance", null);
			ReflectionHelpers.setStaticField(btclass, "sHandler", null);
		}

	// https://github.com/robolectric/robolectric/issues/2068
        Class clazz = ReflectionHelpers.loadClass(getClass().getClassLoader(), "android.view.WindowManagerGlobal");
        Object instance = ReflectionHelpers.callStaticMethod(clazz, "getInstance");

        // We essentially duplicate what's in {@link WindowManagerGlobal#closeAll} with what's below.
        // The closeAll method has a bit of a bug where it's iterating through the "roots" but
        // bases the number of objects to iterate through by the number of "views." This can result in
        // an {@link java.lang.IndexOutOfBoundsException} being thrown.
        Object lock = ReflectionHelpers.getField(instance, "mLock");

        ArrayList<Object> roots = ReflectionHelpers.getField(instance, "mRoots");
        //noinspection SynchronizationOnLocalVariableOrMethodParameter
        synchronized (lock) {
            for (int i = 0; i < roots.size(); i++) {
                ReflectionHelpers.callInstanceMethod(instance, "removeViewLocked",
                        ReflectionHelpers.ClassParameter.from(int.class, i),
                        ReflectionHelpers.ClassParameter.from(boolean.class, false));
            }
        }

        // Views will still be held by this array. We need to clear it out to ensure
        // everything is released.
        Collection<View> dyingViews = ReflectionHelpers.getField(instance, "mDyingViews");
        dyingViews.clear();

**Comment by intirix** [comment]
I got passed my issue.  Some of my tests validate that an action will take you from one activity to another.  For most tests, I peek at the Intent and assert the component's class name.  For some tests, I actually start the next activity to verify that the payload inside of the intent is correct.  Here is my code:

	public <T extends Activity> ActivityController<T> startNextActivity(ActivityController<?> controller, Class<T> klass) {
		
		final Intent intent = Shadows.shadowOf(controller.get()).getNextStartedActivity();
		
		Assert.assertNotNull("Trying to start next activity, but no activity is starting", intent);
		Assert.assertEquals("Trying to start wrong activity", klass.getCanonicalName(), intent.getComponent().getClassName());
		
		final ActivityController<T> nextController = Robolectric.buildActivity(klass).withIntent(intent);
		nextController.create().start();
		controller.pause().stop().destroy();
		nextController.resume();

		// I had to add a call to WindowManagerGlobal.closeAll()
		
		return nextController;
	}

I don't think this is correct, because I think it is removing roots for the new activity, but memory usage is way down now.

**Comment by jaredsburrows** [comment]
@intirix Nice! So all of your tests pass now?

**Comment by Jatra** [comment]
Hi,

I'm seeing the leaks pretty much the same as JavierSP1209 commented on Aug 19, 2015.

We have a test suite of ~ 2,500 tests, which was running fine with 3.0-rc2, but with 3.0 causing OOM.

I've applied the patches from @intirix in our TestApplication afterTest (implementing TestLifecycleApplication).

but the tests still fail, MAT showing eg:

129,520 instances of "android.app.LoadedApk$ReceiverDispatcher", loaded by "org.robolectric.internal.bytecode.InstrumentingClassLoader @ 0x780c7d880" occupy 314,058,560 (31.78%) bytes. 

373 instances of "org.robolectric.shadows.ShadowApplication$1", loaded by "org.robolectric.internal.bytecode.InstrumentingClassLoader @ 0x780c7d880" occupy 247,285,480 (25.02%) bytes. 


Might there be a way of patching into this?

thanks.

**Comment by PiotrSmietana89** [comment]
Hello

I've encountered this issue while using Roolectric 3.1. Similiar as in other cases mentioned in this thread, heap dump analysis shows that there are very numerous instances of ReceiverDispatcher (in my case about 255k) and hundreds of Activity and ShadowActivity instances, but also:

- ShadowBroadcastReceiver (slightly more instances than ReceiverDispatcher instances)
- IntentReceiverLeaked (the same number of  instances as ReceiverDispatcher)
- LockPatternUtils and AccessibilityManagerService (exactly half the number of ReceiverDispatcher instances)

I wrote about it in more detail on stackoverflow:
http://stackoverflow.com/questions/38833627/robolectric-memory-leak-due-to-numeorus-receiverdispatcher-and-activity-instance

Currently I've settled with forking in build.gradle. It helped, but I'm afraid the issue will return as new test are added.

**Comment by mniami** [comment]
Hi guys, 
do we know why this issue is closed? 
Problem still occurred and it doesn't seem to be solved.
Could you analyze it once more? This could help us a lot with running more than 1500 tests in Android Studio which reaches the limits of our server memory. 
Thanks a lot.

**Comment by jaredsburrows** [comment]
@mniami  https://github.com/robolectric/robolectric/issues/2068#issuecomment-242137794

**Comment by dhaag23** [comment]
Just to follow up on my experience: we have over 5000 tests which we can run in about 3GB. I followed all the activity and fragment cleanup advice and that seemed to help.

**Comment by mniami** [comment]
That could be a true. We have only 1.3GB available memory for our UTs.
Still can't understand why this tests requires 3GB?

On 25 Aug 2016 2:51 a.m., "dirk" <notifications@github.com> wrote:

Just to follow up on my experience: we have over 5000 tests which we can
run in about 3GB. I followed all the activity and fragment cleanup advice
and that seemed to help.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/robolectric/robolectric/issues/1700#issuecomment-242251937>,
or mute the thread

**Comment by jaredsburrows** [comment]
@dhaag23 Can you post an example project or list which comments helped you to help others?

**Comment by dhaag23** [comment]
@jaredsburrows - sorry for not posting details. I think the main thing that helped was rigorously cleaning up activities as per previous suggestions. I use this code:

    public static void destroyFragmentAndActivity(@NonNull Fragment fragment) {
		FragmentActivity activity = fragment.getActivity();
		FragmentManager fragmentManager = activity.getSupportFragmentManager();
		fragmentManager.beginTransaction().remove(fragment).commit();
		fragmentManager.executePendingTransactions();
        activity.finish();
    }

Also, a fair number of memory issues were our own as well, in making sure to stop threads and cleaning up other leaked resources.

**Comment by jaredsburrows** [comment]
@dhaag23 Thanks! I am currently using: https://github.com/robolectric/robolectric/issues/1700#issuecomment-163943815. I will also try adding yours!

**Comment by KevinLeighCrain** [comment]
@dhaag23 Works really well! I placed this in a new teardown method in the base class for unit testing activities and update the solution to check programmatically if there are fragments added to the fragment manager before creating a fragment transaction to remove all added fragments at once in one transaction. Thanks for the post was really helpful and boosts the speed of tests significantly!

**Comment by shadabunique** [comment]
I have  around 2500 unit test cases and I get OOM after around 2000 .
I tried all the work arounds mentioned but no success.
Why is this issue closed ?
